### PR TITLE
IP-483 Use a promise, provide a method for setting callbacks before we execute

### PR DESCRIPTION
**Pull request by danielcooper**
This PR: 
- Uses a promise rather than a future for fgets

### CODE REVIEW

File: lib/routemaster/responses/response_promise.rb [hunk 1]
```diff
@@ -1,16 +1,16 @@
-require 'concurrent/future'
+require 'concurrent/promise'
 require 'concurrent/executor/cached_thread_pool'
 require 'singleton'
 require 'delegate'
 
 module Routemaster
   module Responses
-    class FutureResponse
+    class ResponsePromise < Concurrent::Promise
```
Comment (mezis) [inline lib/routemaster/responses/response_promise.rb]: I'm not comfortable inheriting here. The original code was wrapping the future instead of inheriting; this allowed us to change libraries (from `thread` to `concurrent-ruby`) just recently, because we could guarantee that no on was depending on the interface provided by the other library (just on _our_ interface).

I'd like to do the same here: façade rather than inheritance, and expose jsut the bits we want, through delegation. Presumbably just `on_success` and `on_error`, and/or `then` but with a simpler interface.

In other words, I think we should avoid leaking the details of how we implement promises; both because it give us better maintainability, and removes an opportunity for shooting ourselves in the foot with concurrency issues.

Does that work for you?

**Review by mezis — CHANGES_REQUESTED** [review]

File: lib/routemaster/responses/response_promise.rb (hunk 1 above)
Comment (danielcooper) [reply lib/routemaster/responses/response_promise.rb]: Yeah, that's not unreasonable.

File: spec/routemaster/responses/response_promise_spec.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+  it "can have callbacks set" do
+    passing_spy = spy('passing spy')
+    promise = described_class.new { }
+    success_promise = promise.on_success { passing_spy.on_success }
+    promise.execute
+    promise.value
+    success_promise.value
+    expect(passing_spy).to have_received(:on_success)
+  end
```
Comment (gregbeech) [inline spec/routemaster/responses/response_promise_spec.rb:22]: Might be nice to also test `on_error` just for the sake of completeness.

Comment (gregbeech) [reply spec/routemaster/responses/response_promise_spec.rb:22]: Well, and to make sure it works with our approach of raising exceptions (which I'm not 100% sure it will...)

**Review by gregbeech — COMMENTED** [review]
Generally looks good -- I'd like to see a test for the on_error callback though to make sure that value raising an error doesn't prevent that from running.

File: spec/routemaster/responses/response_promise_spec.rb (hunk 2 above)
Comment (danielcooper) [reply spec/routemaster/responses/response_promise_spec.rb:22]: I've added a test to that effect in api_client_spec.rb

File: spec/routemaster/api_client_spec.rb [hunk 3]
```diff
@@ -102,13 +102,29 @@ def dummy; true; end
...
         it "calls on_success" do
           expect(subject.status).to eq 200
           callback.value #We need to wait before testing if the spy was called
           expect(callback_spy).to have_received(:success)
+          expect(callback_spy).not_to have_received(:error)
+
+        end
+      end
+
+      context "when successful" do
```
Comment (gregbeech) [inline spec/routemaster/api_client_spec.rb]: "when not successful"

File: spec/routemaster/api_client_spec.rb [hunk 4]
```diff
@@ -102,13 +102,29 @@ def dummy; true; end
...
+        let(:status){ 500 }
+        it "calls on_success" do
```
Comment (gregbeech) [inline spec/routemaster/api_client_spec.rb]: calls "on_error"

**Review by gregbeech — APPROVED** [review]
Couple of small comments on the test wording (you've copied a test but not changed the description) but other than that :shipit:

**Review by mezis — APPROVED** [review]
Looks all good!
